runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryPoolAssigner.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.nemo.runtime.executor.data;
+
+import com.google.common.annotations.VisibleForTesting;
+import net.jcip.annotations.ThreadSafe;
+import org.apache.reef.tang.annotations.Parameter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.apache.nemo.conf.JobConf;
+
+import javax.inject.Inject;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+/**
+ * The MemoryPoolAssigner assigns the memory that Nemo uses for writing data blocks from the {@link MemoryPool}.
+ * Memory is represented in chunks of equal size. Consumers of off-heap memory acquire the memory by requesting
+ * a number of {@link MemoryChunk} they need.
+ *
+ * MemoryPoolAssigner currently supports allocation of off-heap memory only.
+ *
+ * The MemoryPoolAssigner pre-allocates user-defined amount of memory at the start.
+ * More memory can be allocated on-demand, but if there is no more memory to allocate, MemoryAllocationException
+ * is thrown and the job fails. // TODO #397: Separation of JVM heap region and off-heap memory region
+ */
+@ThreadSafe
+public class MemoryPoolAssigner {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MemoryPoolAssigner.class.getName());
+
+  private final int chunkSize;
+
+  private static final int MIN_CHUNK_SIZE_KB = 4;
+
+  private final MemoryPool memoryPool;
+
+  @Inject
+  public MemoryPoolAssigner(@Parameter(JobConf.MaxOffheapMb.class) final int maxOffheapMb,
+                            @Parameter(JobConf.ChunkSizeKb.class) final int chunkSizeKb) {
+    if (chunkSizeKb < MIN_CHUNK_SIZE_KB) {
+      throw new IllegalArgumentException("Chunk size too small. Minimum chunk size is 4KB");
+    }
+    final long maxNumChunks = (long) maxOffheapMb * 1024 / chunkSizeKb;
+    if (maxNumChunks > Integer.MAX_VALUE) {
```

Comment:
Long.MAX_VALUE?

Reply:
As we track the number of chunks in integer, this check would make sense. Also, if the number of chunks exceeds MAX_INT, it means that the memory size is set to a too large number (which wouldn't be possible).